src/groupeddataframe/grouping.jl

```diff
@@ -412,28 +412,53 @@ true
 """
 Base.keys(gd::GroupedDataFrame) = GroupKeys(gd)
 
-# Index with GroupKey
-function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
-    gd === parent(key) && return gd[getfield(key, :idx)]
+
+# The allowed key types for dictionary-like indexing
+const AbstractGroupKey = Union{GroupKey, Tuple, NamedTuple}
+
+
+# Find the integer index of a group given any supported scalar index type
+_findgroup(gd::GroupedDataFrame, i::Integer) = i
+
+function _findgroup(gd::GroupedDataFrame, key::GroupKey)
+    gd === parent(key) && return getfield(key, :idx)
     throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
 end
 
-# Index with tuple
-function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+function _findgroup(gd::GroupedDataFrame, key::Tuple)
     for i in 1:length(gd)
-        isequal(Tuple(_groupvalues(gd, i)), key) && return gd[i]
+        isequal(Tuple(_groupvalues(gd, i)), key) && return i
     end
     throw(KeyError(key))
 end
 
-# Index with named tuple
-function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+function _findgroup(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
     if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
         throw(KeyError(key))
     end
-    return gd[Tuple(key)]
+    return _findgroup(gd, values(key))
 end
 
+# Dictionary-like indexing
+Base.getindex(gd::GroupedDataFrame, key::AbstractGroupKey) = gd[_findgroup(gd, key)]
+
+# Array indexing with dictionary-like keys
+Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray{T}) where {T<:AbstractGroupKey} = gd[[_findgroup(gd, k) for k in idxs]]
+
+# InvertedIndex
+function Base.getindex(gd::GroupedDataFrame, idx::InvertedIndex{T}) where T
+    selected = trues(length(gd))
+    if T <: AbstractArray
+        selected[[_findgroup(gd, k) for k in idx.skip]] .= false
```

Comment:
not a big thing, but maybe we can do it in a loop to avoid the allocation of the inner array?

Reply:
Also the design here is not optimal because:
1. it allows mixing integers, tuples, named tuples and `GroupKey`s in `idx.skip` that I think we should not allow
2. having `Bool` value of `true` in `idx.skip` (either bare or in a vector) will lead to wrong results (first index will be selected while it should error)

Like in the case of standard `getindex` above we should make sure that the values passed have homogeneous type, an if this type is integer, we should check if no values is `Bool` there.

A typical pattern to handle this (the `Not` case and the standard case above) would be:
1. have "fast" methods in case `eltype` of passed index is well specified (like the case you handle below with eltype being `Bool`)
2. have "catch all" slower method (that would be called only in obscure cases like `Any[1, 2, 3]`) that would check homogeneity of what was passed to it (and make sure no `Bool` slipped in) and then perform the selection similarly to what you do now.